src/general/IWorldEffect.cs

```diff
@@ -93,3 +95,167 @@ public void OnTimePassed(double elapsed, double totalTimePassed)
+        constantCompoundsDissolvedIntake[carbonDioxide] = Constants.PATCH_CONSTANT_CARBON_DIOXYDE_INPUT;
+        constantCompoundsDissolvedIntake[nitrogen] = Constants.PATCH_CONSTANT_NITROGEN_INPUT;
+
+        foreach (var intake in constantCompoundsDissolvedIntake)
+        {
+            if (patch.Biome.Compounds.TryGetValue(intake.Key, out EnvironmentalCompoundProperties compoundValue))
+            {
+                // TODO if capped here, use something to scale production
+                compoundValue.Dissolved = Math.Max(
+                    compoundValue.Dissolved + intake.Value,
+                    Constants.DISSOLVED_MIN);
+                patch.Biome.Compounds[intake.Key] = compoundValue;
+            }
+            else
+            {
+                patch.Biome.Compounds[intake.Key] = new EnvironmentalCompoundProperties
+                {
+                    Amount = 0,
+                    Density = 0,
+                    Dissolved = intake.Value / patch.Volume,
+                };
+            }
+        }
+    }
+
+    /// <summary>
+    ///   Computes the factor of over consumption, i.e. the factor by which the most limiting compound consumption
+    ///   should be diminished not to consume more than there is (including compounds that were just produced).
+    /// </summary>
+    public float ComputeOverConsumptionFactor(Dictionary<Compound, float> compoundsProduced, Patch patch)
+    {
+        var overConsumptionScalingFactor = 1f;
+        foreach (var production in compoundsProduced)
+        {
+            var concentrationInput = production.Value / patch.Volume;
+            if (concentrationInput < 0)
+            {
+                if (patch.Biome.Compounds.TryGetValue(
+                        production.Key, out EnvironmentalCompoundProperties compoundValue))
+                {
+                    if (compoundValue.Dissolved <= Constants.DISSOLVED_MIN)
+                        return 0;
+
+                    overConsumptionScalingFactor = Math.Min(
+                        -(compoundValue.Dissolved - Constants.DISSOLVED_MIN) / concentrationInput,
+                        overConsumptionScalingFactor);
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+        }
+
+        return overConsumptionScalingFactor;
+    }
+
+    /// <summary>
+    ///   Reduce ALL compound production by a scaling factor if one compounds happens to be overproduced.
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     This is a heuristic to prevent production of compounds from non-existing ones.
+    ///     A better yet more complicated approach would be to only limit the processes consuming the target compounds.
+    ///   </para>
+    /// </remarks>
+    public Dictionary<Compound, float> ScaleByOverconsumption(Dictionary<Compound, float> compoundsProduced,
```

Comment:
Shouldn't this return void as this modifies the parameter in-place?

Reply:
I think that for some reason it was not actually doing in place changes, so I used a return instead

Reply:
`struct` would be a value type, but Dictionary is a `class`